src/nodes/RegularNode.ts

```diff
@@ -31,8 +31,9 @@ export class RegularNode extends BaseNode {
   public readonly meta: Readonly<Partial<Dictionary<unknown, SchemaMeta>>>;
   public readonly annotations: Readonly<Partial<Dictionary<unknown, SchemaAnnotations>>>;
   public readonly validations: Readonly<Dictionary<unknown>>;
+  public readonly context: Readonly<{ originalRef?: string }>;
 
-  constructor(public readonly fragment: SchemaFragment) {
+  constructor(public readonly fragment: SchemaFragment, context?: { originalRef?: string }) {
```

Comment:
Hmm, maybe instead of passing `context` we could actually provide the "original fragment"?
This way we could also support combiners that were merged + some other properties that were simply placed next to $ref.